base/pkg/cache.jl

```diff
@@ -28,27 +28,33 @@ function mkcachedir()
     mkdir(cache)
 end
 
-
 function prefetch(pkg::AbstractString, url::AbstractString, sha1s::Vector)
     isdir(".cache") || mkcachedir()
+    #TODO: force switch to https
+    #url = LibGit2.normalize_url(url)
     cache = path(pkg)
-    if !isdir(cache)
+    repo = if isdir(cache)
+        LibGit2.GitRepo(cache) # open repo, free it at the end
+    else
         info("Cloning cache of $pkg from $url")
-        try Git.run(`clone -q --mirror $url $cache`)
-        catch
-            rm(cache, recursive=true)
-            rethrow()
+        try
+            # clone repo, free it at the end
+            LibGit2.clone(url, cache, bare = true, remote_cb = LibGit2.mirror_cb)
+        catch err
+            isdir(cache) && rm(cache, recursive=true)
+            throw(PkgError("Cannot clone $pkg from $url\n", err))
         end
     end
-    Git.set_remote_url(url, dir=cache)
-    in_cache = Git.iscommit(sha1s, dir=cache)
-    if !all(in_cache)
-        info("Updating cache of $pkg...")
-        Git.success(`remote update`, dir=cache) ||
-        error("couldn't update $cache using `git remote update`")
-        in_cache = Git.iscommit(sha1s, dir=cache)
+    try
+        LibGit2.set_remote_url(repo, url)
+        if !all(sha1->LibGit2.iscommit(sha1, repo), sha1s)
+            info("Updating cache of $pkg...")
+            LibGit2.fetch(repo)
+        end
+        filter(sha1->!LibGit2.iscommit(sha1, repo), sha1s)
```

Comment:
This code seems to potentially run `LibGit2.iscommit` twice on all the `sha1s`.   Maybe first run `iscommits = [LibGit2.iscommit(sha1, repo) for sha1 in sha1s]` and then use `!all(iscommits)` and `sha1s[iscommits]`?